services/venice-server/src/main/java/com/linkedin/venice/listener/StorageReadRequestsHandler.java

```diff
@@ -467,8 +467,8 @@ private CompletableFuture<ReadResponse> handleMultiGetRequestInParallel(
             try {
               responseWrapper.addRecord(record);
 
-              if (optionalValueList.isPresent()) {
-                optionalValueList.get().add(record.value.remaining());
+              if (responseValueList != null) {
+                responseValueList.add(record.value.remaining());
```

Comment:
Something else which I thought was interesting last time I looked at this code is that if we set by index rather than using `add()` then we could get rid of the locking... The `responseKeyList.add` is probably unsafe the way it is now (including before your change).

Reply:
Yes, with async processing, its not right to call `add`. But since its just list of value sizes which is probably fine if they are out of order.

Reply:
Bht `ArrayList` is not thread safe... so its `count` could get an incorrect value...?

Reply:
You mean `IntList` ?  If I do add(index, value) that should be safe. We set the list when `CompletableFuture.allOf` so count should be called only after that ?

Reply:
`add(index, value)` should be thread safe, from my understanding (as long as there is not resizing... which is true, in our case).